raiden/network/proxies/user_deposit.py

```diff
@@ -236,118 +256,204 @@ def _init(
-                        f"With the new deposit of {amount_to_deposit}, the deposit "
-                        f"limit of {whole_balance_limit} would be exceeded."
-                    )
-                    raise BrokenPreconditionError(msg)
-
-                if total_deposit <= previous_total_deposit:
-                    msg = (
-                        f"Current total deposit {previous_total_deposit} is already larger "
-                        f"than the requested total deposit amount {total_deposit}"
-                    )
-                    raise BrokenPreconditionError(msg)
-
-                if current_balance < amount_to_deposit:
-                    msg = (
-                        f"new_total_deposit - previous_total_deposit = {amount_to_deposit} "
-                        f"can not be larger than the available balance {current_balance}, "
-                        f"for token at address {to_checksum_address(token.address)}"
-                    )
-                    raise BrokenPreconditionError(msg)
+        previous_total_deposit, amount_to_deposit = self._deposit_preconditions(
+            beneficiary, total_deposit, given_block_identifier, token
+        )
 
+        # The call to `_deposit_preconditions` makes sure the deposit was valid
+        # at block `given_block_identifier`, the check below is for another
+        # concurrent deposits, with a higher value. Since concurrent calls is
+        # not an error, just a race condition, just wait for the result of the
+        # other operation.
+        current_inflight = self._inflight_deposits.get(beneficiary)
+        if current_inflight is not None and current_inflight.total_deposit >= total_deposit:
+            current_inflight.async_result.get()
+            return
+
+        with self._deposit_inflight(beneficiary, total_deposit):
             log_details = {
                 "given_block_identifier": format_block_id(given_block_identifier),
                 "previous_total_deposit": previous_total_deposit,
             }
-            self._deposit(
-                beneficiary=beneficiary,
-                token=token,
-                total_deposit=total_deposit,
-                amount_to_deposit=amount_to_deposit,
-                log_details=log_details,
+            estimated_transaction = self.client.estimate_gas(
+                self.proxy, "deposit", log_details, beneficiary, total_deposit
             )
 
-    def effective_balance(self, address: Address, block_identifier: BlockSpecification) -> Balance:
-        """ The user's balance with planned withdrawals deducted. """
-        balance = self.proxy.functions.effectiveBalance(address).call(
-            block_identifier=block_identifier
+            if estimated_transaction is not None:
+                transaction_hash = self.client.transact(estimated_transaction)
+
+            self._deposit_check_result(transaction_hash, token, total_deposit, amount_to_deposit)
```

Comment:
Isn't `transaction_hash` undefined when `estimated_transaction is None`?

Reply:
Yes. The first argument of `_deposit_check_result` is an `Optional`. I didn't want to change much of the code now, just reuse what was there.